src/Shared/RoslynUtils/SymbolExtensions.cs

```diff
@@ -121,6 +146,61 @@ public static bool TryGetAttributeImplementingInterface(this ImmutableArray<Attr
         return false;
     }
 
+    public static bool TryGetAttributeImplementingInterface(this ImmutableArray<AttributeData> attributes, string[] interfaceName, [NotNullWhen(true)] out AttributeData? matchedAttribute)
+    {
+        foreach (var attributeData in attributes)
+        {
+            if (attributeData.AttributeClass is not null && attributeData.AttributeClass.Implements(interfaceName))
+            {
+                matchedAttribute = attributeData;
+                return true;
+            }
+        }
+
+        matchedAttribute = null;
+        return false;
+    }
+
+    public static bool Implements(this ITypeSymbol type, params string[] interfaceName)
```

Comment:
do you need the params on this and the method below since you seem to only pass in single arrays? not sure whether it affects generated code.
(I'm assuming you copied this from someplace where they need a more general version)

Reply:
by the way, it seems surprising that these useful methods aren't exposed in code analysis itself?

Reply:
Roslyn usually uses pattern matching when testing for types instead of arrays. For example, https://github.com/dotnet/roslyn/blob/dd7c29de3e172a3abf8768a493d925d54f9a15a9/src/Features/Core/Portable/EmbeddedLanguages/EmbeddedLanguageDetector.cs#L338-L357

I imagine it is faster, but it would be a method per type which would be very verbose.